R/getInsertSizeDistFromBam.R

```diff
@@ -0,0 +1,108 @@
+#' @title Tabulate insert sizes from paired-end alignments in bam files.
+#'
+#' @description Read and tabulate the insert sizes from paired-end alignments
+#'   contained in one or several bam files. By default, all properly aligned
+#'   read pairs are included. Optionally, alignments can be restricted to
+#'   those in a specific genomic region region (\code{regions} argument) or
+#'   the number of alignments read can be limited (\code{nmax} argument).
+#'
+#' @author Michael Stadler
+#'
+#' @param fname \code{character} vector with one or several bam files. If
+#'   multiple files are given, insert sizes from all will be pooled and
+#'   tabulated together.
+#' @param regions \code{GRanges} object. Only alignments falling into these
+#'   regions will be used. If \code{NULL} (the default), all alignments are
+#'   used.
+#' @param nmax \code{numeric(1)} specifying the maximal number of alignments to
+#'   read. If \code{NA} (the default), the alignments in \code{regions} (if
+#'   \code{regions} are not \code{NULL}) or in the bam file will be used.
+#' @param isizemax \code{numeric(1)} specifying the maximal insert size to
+#'   report. Larger insert sizes will be set to \code{isizemax} with a warning.
+#'
+#' @return \code{integer} vector with the number of insert sizes. The element at
+#'   position \code{i} gives the observed number of alignment pairs with an
+#'   insert size of \code{i}. The number of insert sizes greater than
+#'   \code{isizemax} that were set to \code{isizemax} are reported in the
+#'   attribute \code{"ncapped"}.
+#'
+#' @seealso \code{\link[Rsamtools]{scanBam}} used to read alignments.
+#'
+#' @examples
+#' if (requireNamespace("Rsamtools", quietly = TRUE)) {
+#'     bamf <- system.file("extdata", "getInsertSizeDistFromBam", "atac_mm10.bam",
+#'                         package = "swissknife")
+#'     isize <- getInsertSizeDistFromBam(bamf)
+#'     attr(isize, "ncapped")
+#'     plot(isize, type = "l",
+#'          xlab = "Insert size (bp)", ylab = "Number of fragments")
+#' }
+#'
+#' @importFrom IRanges IRangesList
+#' 
+#' @export
+getInsertSizeDistFromBam <- function(fname,
+                                     regions = NULL,
+                                     nmax = NA_integer_,
+                                     isizemax = 800) {
```

Comment:
would it be possible to have a parameter like `exclude = "chrM"` to more easily exclude chromosomes that are not of interest, in case a GRanges isn't available. The user could also make a GRanges object without chrM, but if the exclude parameter is easy to implement this may be quicker.

Reply:
I agree that this would be very convenient and probably a common use-case. What would you suggest as default value? `exclude = "chrM"`?

Reply:
Yes, or `exclude=c("chrM", "chrY", "chrX)`

Reply:
Ok, I have added `exclude = c("chrM", "chrY", "chrX")` - it will be ignored if `regions` is set, but construct `regions` on the fly using the chromosome length information from the bam header after filtering out the chromosomes in `exclude`.